### DISCUSSION TITLE
What's the recommended way to run a script at build time?

### DISCUSSION THREAD

**Discussion by danptr** [vercel/next.js#17479] [Help]
Probably by changing the webpack config?

**Comment by jamesmosier** [comment]
Hi @Relypete. What is the script doing?

**Reply by danptr** [reply]
A script to generate a static XML file which will be served through the `/public` folder.

**Reply by jamesmosier** [reply]
Could you just run it before `next build`? 

`"build": "node generate-xml.js && next build"`

Otherwise, as you mentioned, you could use webpack and execute your script before the build via a plugin [like this](https://github.com/hashicorp/prebuild-webpack-plugin)

**Comment by DeBraid** [comment]
TIL, you can run nodejs scripts at build time with `webpack` config.  Below, in the context of sitemap generation:

> Finally, override your `next.config.js` to run this script as part of the build process. Your generated file gets created at public/sitemap.xml which then gets moved to the root of your site (requires Next v9.0+).

```
module.exports = {
  webpack: (config, { isServer }) => {
    if (isServer) {
      require('./scripts/generate-sitemap');
    }

    return config;
  }
};
```

via https://leerob.io/blog/nextjs-sitemap-robots cc @leerob

**Reply by sebastienbarre** [reply]
Unfortunately, the webpack/isServer approach this doesn't work if you use `next.config.mjs`...

**Reply by leerob** [reply]
You can use a postbuild script: https://github.com/leerob/leerob.io/blob/main/package.json#L6

**Reply by sebastienbarre** [reply]
Thanks @leerob ; I ended using [iamvishnusankar/next-sitemap](https://github.com/iamvishnusankar/next-sitemap), easy peasy.

**Reply by scottgrunerud** [reply]
@leerob Are postbuild scripts able read .env vars at build time? I have a .env file and have: `postbuild: 'npx ts-node ./my-script'` in my  package.json but it complains that .env vars are undefined ?

**Reply by deiga** [reply]
@scottgrunerud Only if the called script has the ability to do so. AFAIK neither npx nor ts-node read `.env` files automatically without instructing them to

**Reply by mihaibirsan** [reply]
> Unfortunately, the webpack/isServer approach this doesn't work if you use `next.config.mjs`...

For what it's worth, I was able to use `import('./scripts/generate-sitemap.js')` and it worked.

**Comment by MauriceWebb** [comment]
@danptr 

For any prebuild steps, just add the `"prebuild"` property to your  `package.json`'s `"scripts"` property:

```json
{
  ...
  "scripts": {
    "build": "next build",
    "postbuild": "node ./path-to/my-prebuild-script.mjs",
    "dev": "next dev",
    "start": "next start",
    "lint": "next lint"
  }
  ...
}
```

When Next.js goes to run the `"build"` script, the `"prebuild"` script will get executed first and once completed, the `"build"` script will execute.

As @DeBraid referenced, checkout @leerob 's tutorial on creating a dynamic sitemap. [In his example](https://leerob.io/blog/nextjs-sitemap-robots#dynamic-sitemaps), he adds the script to the `"postbuild"` script.

**Reply by mitchierichie** [reply]
is this documented anywhere? I can't find it

**Reply by SimplyComplexable** [reply]
@mitchierichie  https://docs.npmjs.com/cli/v6/using-npm/scripts#pre--post-scripts

**Reply by mitchierichie** [reply]
> @mitchierichie  https://docs.npmjs.com/cli/v6/using-npm/scripts#pre--post-scripts

Hey, thanks. 😅

**Reply by kcrwfrd** [reply]
The formerly helpful link to a tutorial on Lee Rob's site utilizing a `postbuild` script now redirects. Here's an archived version of the tutorial: https://web.archive.org/web/20231206223316/https://leerob.io/blog/nextjs-sitemap-robots#dynamic-sitemaps

**Comment by janigowski** [comment]
You may also be interested in this tool to fetch & save data in build time:
https://www.npmjs.com/package/next-plugin-preval

Use cases:

> The primary mechanism Next.js provides for static data is getStaticProps — which is a great feature and is the right tool for many use cases. However, there are other use cases for static data that are not covered by getStaticProps.
> 
> Site-wide data: if you have static data that's required across many different pages, getStaticProps is a somewhat awkward mechanism because for each new page, you'll have to re-fetch that same static data. For example, if you use getStaticProps to fetch content for your header, that data will be re-fetched on every page change.
> 
> Static data for API routes: It can be useful to pre-evaluate data fetches in API routes to speed up response times and offload work from your database. getStaticProps does not work for API routes while next-plugin-preval does.
> 
> De-duped and code split data: Since next-plugin-preval behaves like importing JSON, you can leverage the optimizations bundlers have for importing standard static assets. This includes standard code-splitting and de-duping.
> 
> Zero runtime: Preval files don't get sent to the browser, only their outputted JSON.

**Reply by SimplyComplexable** [reply]
This is really cool, but is there a way to run scripts within the project? I'm using a Typescript project with various dependencies and the script I need to run is written in that architecture. I've tried using ts-node to run it, but keep running into issues. I really just want to use the final webpack config from the project to run the script, but I can't figure out a way to do that. Any suggestions?

**Reply by shamoons** [reply]
Same - would love to keep it all consolidated
